File: backend/src/middlewares/authenticate.js

import jwt from "jsonwebtoken";
import bcrypt from "bcrypt"; // Thêm dòng này
import pool from "../db.js";
import ldap from "ldapjs";
import dotenv from "dotenv";
dotenv.config();

const SECRET = process.env.JWT_SECRET;

// Hàm kết nối và kiểm tra LDAP
export const authenticateLDAP = (username, password) => {
    return new Promise((resolve, reject) => {
        const adminClient = ldap.createClient({ url: "ldap://localhost:389" });

        adminClient.bind("cn=admin,dc=example,dc=com", "admin", (err) => {
            if (err) return reject(new Error("LDAP admin bind failed"));

            const searchOptions = {
                scope: "sub",
                filter: `(uid=${username})`,
            };

            adminClient.search("dc=example,dc=com", searchOptions, (err, result) => {
                if (err) return reject(new Error("LDAP search error"));

                let userDN = null;

                result.on("searchEntry", (entry) => {
                    userDN = entry.objectName.toString();
                });

                result.on("end", () => {
                    adminClient.unbind();

                    if (!userDN) return reject(new Error("User not found"));

                    const userClient = ldap.createClient({ url: "ldap://localhost:389" });

                    userClient.bind(userDN, password, (err) => {
                        userClient.unbind();
                        if (err) {
                            return reject(new Error("Invalid LDAP credentials"));
                        } else {
                            return resolve(userDN); // Thành công
                        }
                    });
                });
            });
        });
    });
};

// Hàm lấy menu của user từ DB
const getMenusByUserId = async (userId) => {
    const result = await pool.query(`
      SELECT DISTINCT m.id, m.name, m.path
      FROM users u
      JOIN user_groups ug ON u.id = ug.user_id
      JOIN group_roles gr ON ug.group_id = gr.group_id
      JOIN role_menus rm ON gr.role_id = rm.role_id
      JOIN menus m ON rm.menu_id = m.id
      WHERE u.id = $1
    `, [userId]);

    return result.rows; // [{id, name, path}]
};


// Middleware: xử lý login hỗn hợp
export const login = async (req, res) => {
    const { username, password } = req.body;
    if (!username || !password) {
        return res.status(400).json({ error: "Username và password là bắt buộc" });
    }
    // Kiểm tra định dạng username
    const usernameRegex = /^[a-zA-Z0-9_]+$/;
    if (!usernameRegex.test(username)) {
        return res.status(400).json({ error: "Username không hợp lệ. Chỉ cho phép chữ cái, số và dấu gạch dưới." });
    }
    // Kiểm tra định dạng password
    const passwordRegex = /^(?=.*[!@#$%^&*()_+{}\[\]:;<>,.?~\\/-]).{8,}$/;
    if (!passwordRegex.test(password)) {
        return res.status(400).json({ error: "Mật khẩu phải từ 8 ký tự trở lên và chứa ít nhất 1 ký tự đặc biệt." });
    }

    try {
        // Check user trong DB
        const userRes = await pool.query("SELECT * FROM users WHERE username = $1", [username]);
        const user = userRes.rows[0];

        // Nếu có user và có password trong DB → kiểm tra bằng bcrypt
        if (user && user.password) {
            const match = await bcrypt.compare(password, user.password);

            if (match) {
                const token = jwt.sign({ id: user.id, username: user.username, avatar: user?.avatar, email: user?.email }, SECRET, { expiresIn: "1h" });
                const menus = await getMenusByUserId(user.id);
             
                return res.json({ message: "Đăng nhập thành công", token, menus });
            }
        }

        // Nếu không login được → thử LDAP
        await authenticateLDAP(username, password);

        // Nếu login LDAP thành công → thêm user nếu chưa có
        let userId = user?.id;

        if (!userId) {
            const insertRes = await pool.query(
                "INSERT INTO users (username, name, password) VALUES ($1, $2, $3) RETURNING id",
                [username, username, ""] // Không lưu password từ LDAP
            );
            userId = insertRes.rows[0].id;
        }

        const token = jwt.sign({ id: userId, username, avatar: user?.avatar }, SECRET, { expiresIn: "1h" });
        const menus = await getMenusByUserId(userId);
        return res.json({ message: "Đăng nhập thành công LDAP", token, menus });
    } catch (err) {
        console.error("Login error:", err);
        return res.status(401).json({ error: "Đăng nhập thất bại, sai tài khoản hoặc mật khẩu" });
    }
};

export const authMiddleware = (req, res, next) => {
    const authHeader = req.headers["authorization"];
    const token = authHeader && authHeader.split(" ")[1]; // Bearer <token>

    if (!token) return res.status(401).json({ error: "Token không tồn tại" });

    try {
        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        req.user = decoded; // lưu thông tin vào req để dùng tiếp
        next();
    } catch (err) {
        return res.status(403).json({ error: "Token không hợp lệ hoặc đã hết hạn" });
    }
};

File: backend/app/middlewares/authenticate.py
# app/utils/authenticate.py
import jwt
import bcrypt
import psycopg2
from flask import request, jsonify, g
from datetime import datetime, timedelta
import os
from dotenv import load_dotenv
from db import get_connection, release_connection

load_dotenv()

JWT_SECRET = os.getenv("JWT_SECRET")
JWT_EXPIRES_IN = 60 * 60 * 3  # seconds (3 hours)




# Get menus from user id
def get_menus_by_user_id(user_id):
    conn = get_connection()
    if not conn:
        print("Could not get DB connection")
        return []

    try:
        cur = conn.cursor()
        cur.execute("""
            SELECT DISTINCT m.id, m.name, m.path
            FROM users u
            JOIN user_groups ug ON u.id = ug.user_id
            JOIN group_roles gr ON ug.group_id = gr.group_id
            JOIN role_menus rm ON gr.role_id = rm.role_id
            JOIN menus m ON rm.menu_id = m.id
            WHERE u.id = %s
        """, (user_id,))
        rows = cur.fetchall()
        return [{"id": r[0], "name": r[1], "path": r[2]} for r in rows]
    except Exception as e:
        print(f"Error fetching menus: {e}")
        return []
    finally:
        cur.close()
        release_connection(conn)



# Hybrid login
def login():
    data = request.get_json()
    username = data.get("username")
    password = data.get("password")

    if not username or not password:
        return jsonify({"error": "Username và password là bắt buộc"}), 400

    if not username.isalnum():
        return jsonify({"error": "Username không hợp lệ. Chỉ cho phép chữ cái và số."}), 400

    if len(password) < 8 or not any(c in "!@#$%^&*()_+-=[]{}|;:,.<>?" for c in password):
        return jsonify({"error": "Mật khẩu phải từ 8 ký tự và có ít nhất 1 ký tự đặc biệt."}), 400

    conn = get_connection()
    cur = conn.cursor()
    cur.execute("SELECT * FROM users WHERE username = %s", (username,))
    user = cur.fetchone()
    print(f"User found: {user}")

    if user:
        hashed_pw = user[2].encode('utf-8')  # fix: dùng user[2] thay vì user[3]
        if bcrypt.checkpw(password.encode('utf-8'), hashed_pw):
            payload = {
                "id": user[0],
                "username": user[1],
                "email": user[5],       # fix nếu cần
                "avatar": user[4],
                "exp": datetime.utcnow() + timedelta(seconds=JWT_EXPIRES_IN)
            }
            token = jwt.encode(payload, JWT_SECRET, algorithm="HS256")
            menus = get_menus_by_user_id(user[0])
            return jsonify({
                "message": "Đăng nhập thành công",
                "token": token,
                "menus": menus
            })

    return jsonify({"error": "Sai tài khoản hoặc mật khẩu"}), 401


# Middleware
def auth_middleware(f):
    def wrapper(*args, **kwargs):
        auth_header = request.headers.get("Authorization")
        if not auth_header or not auth_header.startswith("Bearer "):
            return jsonify({"error": "Token không tồn tại"}), 401

        token = auth_header.split(" ")[1]
        try:
            decoded = jwt.decode(token, JWT_SECRET, algorithms=["HS256"])
            g.user = decoded  # attach to Flask global context
        except jwt.ExpiredSignatureError:
            return jsonify({"error": "Token đã hết hạn"}), 403
        except jwt.InvalidTokenError:
            return jsonify({"error": "Token không hợp lệ"}), 403

        return f(*args, **kwargs)
    wrapper.__name__ = f.__name__  # Preserve function name
    return wrapper

# Middleware kiểm tra API Key
def require_api_key(f):
    from functools import wraps

    @wraps(f)
    def decorated(*args, **kwargs):
        api_key = request.headers.get("X-API-KEY")
        expected_key = os.getenv("ADMIN_API_KEY")

        if not api_key or api_key != expected_key:
            return jsonify({"error": "Unauthorized: API Key không hợp lệ"}), 401

        return f(*args, **kwargs)

    return decorated
